src/Mod/Arch/ArchRoof.py

```diff
@@ -19,18 +19,28 @@
 #*                                                                         *
 #***************************************************************************
 
-import FreeCAD, Draft, math, ArchComponent, DraftVecUtils
+import math
+
+import ArchComponent
+import Arch_rc
+import Draft
+import DraftGeomUtils
+import DraftVecUtils
+import FreeCAD
+import Part
```

Comment:
The reason Yorik imports many modules inside of methods is because importing `Part` is an expensive call, so he prefers to delay the import until the module is actually used. `DraftGeomUtils` also uses `Part` directly so for these two cases I recommend delaying the import by using the `lazy_loader` module. This module was introduced by the Path Workbench but is available for use by everybody.

You can see examples inside the submodules of `DraftGeomUtils`.

```py
import lazy_loader.lazy_loader as lz

Part = lz.LazyLoader("Part", globals(), "Part")
DraftGeomUtils = lz.LazyLoader("DraftGeomUtils", globals(), "DraftGeomUtils")
```

On the code that I am refactoring, I always try to delay the import of these two modules. Without proper profiling it's hard to know where the slow parts are but for the time being I think it's a good solution. For example, the Path guys did some profiling of their code, and they delayed the import of `Draft` because of it.

Reply:
I do not know anything about this. When/if Yorik merges he will no doubt address this.

The main goal of this update is to fix several (sometimes hard to detect bugs). I am sure that there are many cosmetic issues with the file, but IMO they can be addressed later (although I have tried to improve things already).

There is still a ToDo list:

1. Rename arguments and properties.
2. Get rid of the Flip property, the code should handle this automatically.
3. Add the option to work down from the base so that any ridges caused be uneven angles and thicknesses are at the bottom of the roof.
4. Make the creation of solids more efficient by extruding a 'tilted' face instead of using a 'common' Boolean operation.
5. Check if a roof can be create from a base wire with a normal that is not parallel to the global Z.
6. And indeed: Comply with coding guide lines and other cosmetics.

The main algorithm has its obvious limitations. The outline of a segment is only based on its direct neighbors. So before investing more time in the code it may also be wise the step back and reconsider things.

Reply:
I imagine that most commands in the Arch WB rely on Part and DraftGeomUtils so maybe the suggestion to use the lazy_loader makes less sense here.

Reply:
Actually it's the opposite. Since most commands on Arch depend on Part and DraftGeomUtils, these two modules should be delayed in import in every module. Why? Because when the Arch Workbench loads, it loads all modules (maybe including Part and DraftGeomUtils), making the loading slow, which is what Yorik wants to avoid.

Arch already loads pretty slow when compared to, say, Draft. I think something can be improved here; maybe it's the reading of the IFC files or something like that, but there is something definitely slow in Arch when it starts.

Reply:
So if I understand you correctly, the splitting up of python files into many smaller files, is a contributing factor here. Part and DraftGeomUtils are not imported once for the whole WB, but for each file separately.

Reply:
Negative. Python imports a module only once, then the module is cached, so it isn't imported more than once. Exactly what is making Arch slow on start, I am not sure. But if we delay the import of `Part`, we at least remove this as a cause.